File: payment/src/main/java/io/axoniq/demo/bikerental/payment/PaymentController.java
package io.axoniq.demo.bikerental.payment;

import io.axoniq.demo.bikerental.coreapi.payment.ConfirmPaymentCommand;
import io.axoniq.demo.bikerental.coreapi.payment.PaymentStatus;
import io.axoniq.demo.bikerental.coreapi.payment.PaymentStatusNamedQueries;
import io.axoniq.demo.bikerental.coreapi.payment.RejectPaymentCommand;
import io.axoniq.demo.bikerental.payment.support.CommandDispatcher;
import io.axoniq.demo.bikerental.payment.support.QueryDispatcher;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.concurrent.CompletableFuture;

@RestController
public class PaymentController {

    private static final String GET_STATUS = "getStatus";

    private final QueryDispatcher queryDispatcher;
    private final CommandDispatcher commandDispatcher;

    public PaymentController(QueryDispatcher queryDispatcher, CommandDispatcher commandDispatcher) {
        this.queryDispatcher = queryDispatcher;
        this.commandDispatcher = commandDispatcher;
    }

    @GetMapping("/status/{paymentId}")
    public CompletableFuture<PaymentStatus> getStatus(@PathVariable("paymentId") String paymentId) {
        return queryDispatcher.query(GET_STATUS, paymentId, PaymentStatus.class);
    }

    @GetMapping("/findPayment")
    public CompletableFuture<String> findPaymentId(@RequestParam("reference") String paymentReference) {
        return queryDispatcher.query(PaymentStatusNamedQueries.GET_PAYMENT_ID, paymentReference, String.class);
    }

    @PostMapping("/acceptPayment")
    public CompletableFuture<Void> confirmPayment(@RequestParam("id") String paymentId) {
        return commandDispatcher.send(new ConfirmPaymentCommand(paymentId));
    }

    @PostMapping("/rejectPayment")
    public CompletableFuture<Void> rejectPayment(@RequestParam("id") String paymentId) {
        return commandDispatcher.send(new RejectPaymentCommand(paymentId));
    }

    @GetMapping("/status")
    public Flux<PaymentStatus> getStatus(@RequestParam(value = "status", required = false) PaymentStatus.Status status) {
        return Mono.fromFuture(queryDispatcher.queryMany(PaymentStatusNamedQueries.GET_ALL_PAYMENTS,
                                                         status, PaymentStatus.class))
                   .flatMapMany(Flux::fromIterable);
    }
}

File: payment/src/payment-controller.ts
import { Router } from 'express';
import {
  CommandDispatcher,
  ConfirmPaymentCommand,
  JavaTypes,
  PaymentStatus,
  PaymentStatusNamedQueries,
  PaymentStatusValue,
  QueryDispatcher,
  RejectPaymentCommand,
} from '@bikerental/core-api';

const GET_STATUS = 'getStatus';

export function paymentRouter(
  queryDispatcher: QueryDispatcher,
  commandDispatcher: CommandDispatcher,
): Router {
  const router = Router();

  router.get('/status/:paymentId', async (req, res) => {
    res.json(await queryDispatcher.query<PaymentStatus>(GET_STATUS, req.params.paymentId, PaymentStatus.TYPE));
  });

  router.get('/findPayment', async (req, res) => {
    const paymentId = await queryDispatcher.query<string>(
      PaymentStatusNamedQueries.GET_PAYMENT_ID,
      String(req.query.reference ?? ''),
      JavaTypes.STRING,
    );
    res.type('text/plain').send(paymentId ?? '');
  });

  router.post('/acceptPayment', async (req, res) => {
    await commandDispatcher.send(new ConfirmPaymentCommand(String(req.query.id ?? '')));
    res.status(200).end();
  });

  router.post('/rejectPayment', async (req, res) => {
    await commandDispatcher.send(new RejectPaymentCommand(String(req.query.id ?? '')));
    res.status(200).end();
  });

  router.get('/status', async (req, res) => {
    const status = req.query.status ? (String(req.query.status) as PaymentStatusValue) : null;
    res.json(
      await queryDispatcher.queryMany<PaymentStatus>(
        PaymentStatusNamedQueries.GET_ALL_PAYMENTS,
        status,
        PaymentStatus.TYPE,
      ),
    );
  });

  return router;
}
